rfcs/71-write-only-world-query.md

```diff
@@ -0,0 +1,385 @@
+It is not unsound to erronously populate this `FilteredAccess`. It is just used
+as a hint for triggering error messages.
+
+This adds another `FilteredAccess` to the query. This `FilteredAccess`
+is only written-to and read-from at app initialization when setting up the systems.
+
+### Triggering the warning
+
+Consider `ui_layout_system`:
+
+https://github.com/bevyengine/bevy/blob/469a19c290861049ad121aa83d72f830d1ec9b40/crates/bevy_ui/src/layout/mod.rs#L215-L230
+
+```rust
+pub fn ui_layout_system(
+    primary_window: Query<(Entity, &Window), With<PrimaryWindow>>,
+    windows: Query<(Entity, &Window)>,
+    ui_scale: Res<UiScale>,
+    mut scale_factor_events: EventReader<WindowScaleFactorChanged>,
+    mut resize_events: EventReader<bevy_window::WindowResized>,
+    mut ui_surface: ResMut<UiSurface>,
+    root_node_query: Query<Entity, (With<Node>, Without<Parent>)>,
+    style_query: Query<(Entity, Ref<Style>), With<Node>>,
+    mut measure_query: Query<(Entity, &mut ContentSize)>,
+    children_query: Query<(Entity, Ref<Children>), With<Node>>,
+    mut removed_children: RemovedComponents<Children>,
+    mut removed_content_sizes: RemovedComponents<ContentSize>,
+    mut node_transform_query: Query<(Entity, &mut Node, &mut Transform, Option<&Parent>)>,
+    mut removed_nodes: RemovedComponents<Node>,
+) {
+```
+
+We would replace
+
+```rust
+    mut node_transform_query: Query<(Entity, &mut Node, &mut Transform, Option<&Parent>)>,
+```
+
+with the following:
+
+```rust
+    mut node_transform_query: Query<(Entity, &mut Node, WriteOnly<Transform>, Option<&Parent>)>,
+```
+
+Now, let's say Daniel — a bevy user — want to add a screenshake effect to their UI
+(please never do this):
+
+```rust
+fn shake_ui(mut nodes: Query<&mut Transform, With<Node>>) {
+  for tran in &mut nodes {
+      tran.translation += random_offset;
+  }
+}
+```
+
+Daniel adds `shake_ui` to their game to spice things up:
+
+
+```rust
+app.add_plugins(DefaultPlugins::new())
+  .add_systems(Update, shake_ui);
+```
+
+Daniel looks at their game. And *\*sad trombone\** nothing happens.
+The UI stays fixed to the screen as if sticked with ultra glue.
+
+**But lo**! What is this? An error in the bevy log output?
+
+```
+2023-04-20T11:11:11.696969Z ERROR bevy_ecs: [B0006] In `Update`, the system `shake_ui` writes the `Transform`, but `ui_layout_system` in `PostUpdate` later overwrites it unconditionally.
+```
+
+| 👆 **We will refer to this error as `B0006` for the rest of this RFC** |
+|------------------------------------------------------------------------|
+
+Daniel reads the error message, goes to <https://bevyengine.org/learn/errors/#b0006>
+and learns that they should — I quote — "schedule `shake_ui` in `PostUpdate`,
+and label it as `shake_ui.after(ui_layout_system)` to effectively update `Transform`".
+
+Daniel updates their app schedule code:
+
+```rust
+app.add_plugins(DefaultPlugins::new())
+  .add_systems(PostUpdate, shake_ui.after(ui_layout_system));
```

Comment:
This isn't actually "fully correct" for this scenario. Making changes to the `Transform` after `ui_layout_system` does ensure it won't get stomped on by that system. And for the _current Bevy systems written_ I do think it would work. But the computed node Transform is considered an output of the `UiSystem::Layout` set. Anything trying to consume the "final computed layout " of a node should schedule relative to that. Therefore if you are modifying Transform (and affecting final layout), you should put your system in `UiSystem::Layout`.